numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1653,10 +1653,38 @@ _array_fromobject_generic(
     /* fast exit if simple call */
     if (PyArray_CheckExact(op) || (subok && PyArray_Check(op))) {
         oparr = (PyArrayObject *)op;
-        if (type == NULL) {
+        PyArray_Descr* dtype = NULL;
+        /*
+         * Skip AdaptDescriptorToArray if the supplied dtype class does not
+         * match the dtype of the input array or if we have to determine the
+         * descriptor because only the DType was given. This means we avoid
+         * inspecting array values twice for example if someone does:
+         *
+         * >>> arr = np.array(["asdf", "fdsa"], dtype=object)
+         * >>> np.array(arr, dtype="U")
+         */
+        if ((NPY_DTYPE(PyArray_DESCR(oparr)) == dt_info.dtype) ||
+            ((dt_info.descr == NULL) && (dt_info.dtype != NULL))) {
+            dtype = PyArray_AdaptDescriptorToArray(
+                    oparr, dt_info.dtype, dt_info.descr);
```

Comment:
One thing I just realized:  Once we called `PyArray_AdaptDescriptorToArray` we can use the result also outside the fast-path:  it is always correct and avoids calling it a second time just as much.

There is also a `string -> datetime` special path (exactly those two), but it feels like probably less awkward to just do:
```
Py_SETREF(dt_info.descr, dtype);
```

Reply:
This ended up being tricky for me because the `SETREF` mutates the pointer and the mutated pointer needs to be available to the caller to decref. I ended up passing the descriptor by reference so that the mutated pointer is available to the caller.